File: typescript/problem_78.ts
class ListNode78ts { constructor(public val:number, public next:ListNode78ts|null=null){} }

export function mergeKLists(lists: (ListNode78ts|null)[]): ListNode78ts|null {
    const heap: ListNode78ts[] = lists.filter(Boolean) as ListNode78ts[];
    heap.sort((a,b)=>a.val-b.val);
    const dummy = new ListNode78ts(0); let cur = dummy;
    while (heap.length) {
        const node = heap.shift()!; cur.next=node; cur=cur.next;
        if (node.next) {
            let i=0; while(i<heap.length&&heap[i].val<=node.next.val)i++;
            heap.splice(i,0,node.next);
        }
    }
    return dummy.next;
}

function fl78(vals:number[]):ListNode78ts|null{const d=new ListNode78ts(0);let c=d;for(const v of vals){c.next=new ListNode78ts(v);c=c.next;}return d.next;}
function tl78(h:ListNode78ts|null):number[]{const r:number[]=[];while(h){r.push(h.val);h=h.next;}return r;}

export function problem78(): void {
    console.log("Problem 78:");
    if (JSON.stringify(tl78(mergeKLists([fl78([1,4,7]),fl78([2,5,8]),fl78([3,6,9])]))) !==
        JSON.stringify([1,2,3,4,5,6,7,8,9]))
        throw new Error("assertion failed");
    console.log("All Tests Passed");
}

File: rust/problem_78.rs
use std::collections::BinaryHeap;
use std::cmp::Reverse;

type Link78 = Option<Box<ListNode78>>;
pub struct ListNode78 { pub val: i32, pub next: Link78 }

fn from_vec78(v: &[i32]) -> Link78 {
    let mut h = None;
    for &x in v.iter().rev() { let mut n = Box::new(ListNode78{val:x,next:None}); n.next=h; h=Some(n); }
    h
}
fn to_vec78(mut h: Link78) -> Vec<i32> {
    let mut r = vec![]; while let Some(n) = h { r.push(n.val); h = n.next; } r
}

pub fn merge_k_lists(mut lists: Vec<Link78>) -> Link78 {
    let mut heap: BinaryHeap<Reverse<(i32, usize)>> = BinaryHeap::new();
    for (i, l) in lists.iter().enumerate() {
        if let Some(n) = l { heap.push(Reverse((n.val, i))); }
    }
    let mut dummy = Box::new(ListNode78{val:0,next:None});
    let mut cur = &mut dummy as *mut Box<ListNode78>;
    while let Some(Reverse((_, i))) = heap.pop() {
        let node = lists[i].take().unwrap();
        lists[i] = node.next;
        if let Some(n) = &lists[i] { heap.push(Reverse((n.val, i))); }
        unsafe { (*cur).next = Some(node); cur = (*cur).next.as_mut().unwrap(); }
    }
    dummy.next
}

pub fn problem78() {
    println!("Problem 78:");
    let lists = vec![from_vec78(&[1,4,7]), from_vec78(&[2,5,8]), from_vec78(&[3,6,9])];
    assert_eq!(to_vec78(merge_k_lists(lists)), vec![1,2,3,4,5,6,7,8,9]);
    println!("All Tests Passed");
}
